api/build.gradle

```diff
@@ -1,32 +1,74 @@
 description = "reactive-streams"
-def jdkFlow = false
+def isJdk9OrGreater = false
 try {
     Class.forName("java.util.concurrent.Flow")
-    jdkFlow = true
+    isJdk9OrGreater = true
 } catch (ClassNotFoundException cnfe) {
 
 }
 
 sourceSets {
     main {
         java {
-            if (jdkFlow)
-              srcDirs = ['src/main/java', 'src/main/java9']
-            else
-              srcDirs = ['src/main/java']
+            srcDirs = ['src/main/java']
         }
     }
+    if (isJdk9OrGreater) java9 {
+        java {
+            srcDirs = ['src/main/java', 'src/main/java9']
+        }
+    }
+}
+
+configurations {
+  mrjarArtifact
 }
 
 jar {
+    if (isJdk9OrGreater) into('META-INF/versions/9') {
+        from(sourceSets.java9.output) {
+            include '**/FlowAdapters*'
+            include '**/module-info.class'
+        }
+    }
```

Comment:
Amends the JAR with the `META-INF/versions/9` root, from the output of the `sourceSets.java9` compile. `include` is used here to avoid duplicating classes which reside in the root of the module, at JDK 6, which are identical anyway.